File: leetcode/311.sparse-matrix-multiplication.cpp
class Solution {
public:
    vector<vector<int>> multiply(vector<vector<int>>& A, vector<vector<int>>& B) {
        
        if (A.size() == 0 || B.size() == 0)
        {
            return vector<vector<int>>();
        }
        
        vector<vector<pair<int, int>>> sparseB(B.size(), vector<pair<int, int>>());

        for (auto i = 0; i < B.size(); ++i)
        {
            for (auto j = 0; j < B[i].size(); ++j)
            {
                if (B[i][j] != 0){
                    sparseB[i].push_back(make_pair(j, B[i][j]));
                }
            }
        }

        int m = A.size();
        int n = B[0].size();

        vector<vector<int>> res(m, vector<int>(n, 0));
        
        for (auto i = 0; i < m; ++i)
        {
            for (auto k = 0; k < A[i].size(); ++k)
            {
                if (A[i][k] != 0){
                    for (auto p = 0; p < sparseB[k].size(); ++p)
                    {
                        int j = sparseB[k][p].first;
                        int val = sparseB[k][p].second;
                        
                        res[i][j] += A[i][k] * val;
                    }
                }
            }
        }

        
        return res;
    }
};

class Solution2 {
public:
    vector<vector<int>> multiply(vector<vector<int>>& A, vector<vector<int>>& B) {
        
        if (A.size() == 0 || B.size() == 0)
        {
            return vector<vector<int>>();
        }
        
        int m = A.size();
        int n = B[0].size();
        
        int t = A[0].size();
        
        vector<vector<int>> res(m, vector<int>(n, 0));
        
        for (auto i = 0; i < m; ++i)
        {
            for (auto j = 0; j < n; ++j)
            {
                for(auto k = 0; k < t; ++k)
                {
                    res[i][j] += A[i][k] * B[k][j];
                }
            }
        }
        
        return res;
    }
};

File: leetcode/311.sparse-matrix-multiplication.py
class Solution(object):
    def multiply(self, A, B):
        """
        :type A: List[List[int]]
        :type B: List[List[int]]
        :rtype: List[List[int]]
        """
        
        if A is None or B is None or len(A) == 0 or len(B) == 0:
            return [[]]

        m = len(A)
        n = len(B[0])

        res = [[0 for _ in range(n)] for _ in range(m)]
        sparseB = [[] for _ in range(len(B))]

        for i in range(len(B)):
            for j in range(len(B[i])):
                sparseB[i].append((j, B[i][j]))


        for i in range(m):
            for k in range(len(A[i])):
                if A[i][k] != 0:
                    for p in range(len(sparseB[k])):
                        j, val = sparseB[k][p]
                        res[i][j] += A[i][k] * val

        return res